common/src/main/java/io/netty/util/ResourceLeakDetector.java

```diff
@@ -453,54 +490,52 @@ public String toString() {
-        StringBuilder buf = new StringBuilder(4096);
-
-        // Append the hint first if available.
-        if (hint != null) {
-            buf.append("\tHint: ");
-            // Prefer a hint string to a simple string form.
-            if (hint instanceof ResourceLeakHint) {
-                buf.append(((ResourceLeakHint) hint).toHintString());
-            } else {
-                buf.append(hint);
-            }
-            buf.append(NEWLINE);
+    private static final class Record extends Throwable {
+
+        private static final Set<String> STACK_TRACE_ELEMENT_EXCLUSIONS = new HashSet<String>();
+        static {
+            STACK_TRACE_ELEMENT_EXCLUSIONS.add("io.netty.util.ReferenceCountUtil.touch");
+            STACK_TRACE_ELEMENT_EXCLUSIONS.add("io.netty.buffer.AdvancedLeakAwareByteBuf.touch");
+            STACK_TRACE_ELEMENT_EXCLUSIONS.add("io.netty.buffer.AbstractByteBufAllocator.toLeakAwareBuffer");
+            STACK_TRACE_ELEMENT_EXCLUSIONS.add(
+                    "io.netty.buffer.AdvancedLeakAwareByteBuf.recordLeakNonRefCountingOperation");
         }
 
-        // Append the stack trace.
-        StackTraceElement[] array = new Throwable().getStackTrace();
-        for (StackTraceElement e: array) {
-            if (recordsToSkip > 0) {
-                recordsToSkip --;
-            } else {
-                String estr = e.toString();
+        private final String hintString;
+        Record next;
 
-                // Strip the noisy stack trace elements.
-                boolean excluded = false;
-                for (String exclusion: STACK_TRACE_ELEMENT_EXCLUSIONS) {
-                    if (estr.startsWith(exclusion)) {
-                        excluded = true;
-                        break;
-                    }
-                }
+        Record(Object hint) {
+            // This needs to be generated even if toString() is never called as it may change later on.
+            hintString = hint instanceof ResourceLeakHint ? ((ResourceLeakHint) hint).toHintString() : hint.toString();
+        }
+
+        Record() {
+           hintString = null;
+        }
 
-                if (!excluded) {
-                    buf.append('\t');
-                    buf.append(estr);
-                    buf.append(NEWLINE);
+        @Override
+        public String toString() {
+            StringBuilder buf = new StringBuilder(4096);
+            if (hintString != null) {
+                buf.append("\tHint: ").append(hintString).append(NEWLINE);
+            }
+
+            // Append the stack trace.
+            StackTraceElement[] array = getStackTrace();
+            // Skip the first three elements.
+            for (int i = 3; i < array.length; i++) {
+                StackTraceElement element = array[i];
+
+                // Strip the noisy stack trace elements.
+                if (STACK_TRACE_ELEMENT_EXCLUSIONS.contains(element.getClassName() + '.' + element.getMethodName())) {
```

Comment:
Hmm, you could probably do even better:  just interleave the class names and method names in STACK_TRACE_ELEMENT_EXCLUSIONS, and then iterate over it.  

You would be much more likely to get the super fast == check that happens for internned strings, and you wouldn't allocate here at all, even on the stack.   

I think you are using `contains` to work around shading, but if you use `ReferenceCountUtil.getClassName()` when you initialize STACK_TRACE_ELEMENT_EXCLUSIONS, I think you can use equals.

Reply:
@carl-mastrangelo not sure I get what you are suggesting.. Can you give an example ? Also again it may not even matter much as its only called once a leak happens.

Reply:
I have another change to this file, I'll show it after you submit your PR.

Reply:
awesome... Thanks